File: mcp-server/src/adapters/simctl.ts
// Thin wrapper around `xcrun simctl` for device + app lifecycle.
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileP = promisify(execFile);

export interface SimDevice {
  udid: string;
  name: string;
  state: "Booted" | "Shutdown" | "Booting" | "Shutting Down" | string;
  runtime: string;
  deviceTypeIdentifier?: string;
}

export async function listDevices(opts: { availableOnly?: boolean } = {}): Promise<SimDevice[]> {
  const args = ["simctl", "list", "devices", "--json"];
  if (opts.availableOnly !== false) args.push("available");
  const { stdout } = await execFileP("xcrun", args, { maxBuffer: 8 * 1024 * 1024 });
  const parsed = JSON.parse(stdout) as { devices: Record<string, Array<Omit<SimDevice, "runtime">>> };
  const out: SimDevice[] = [];
  for (const [runtime, devices] of Object.entries(parsed.devices)) {
    for (const d of devices) {
      out.push({ ...d, runtime });
    }
  }
  return out;
}

export async function bootDevice(udid: string): Promise<{ alreadyBooted: boolean }> {
  // Check state first — boot is non-idempotent (errors if already booted)
  const devs = await listDevices();
  const target = devs.find((d) => d.udid === udid);
  if (!target) throw new Error(`Device ${udid} not found`);
  if (target.state === "Booted") return { alreadyBooted: true };

  await execFileP("xcrun", ["simctl", "boot", udid]);
  // Wait for boot to settle (services come up)
  await execFileP("xcrun", ["simctl", "bootstatus", udid, "-b"]);
  return { alreadyBooted: false };
}

export async function shutdownDevice(udid: string): Promise<void> {
  await execFileP("xcrun", ["simctl", "shutdown", udid]);
}

export async function takeScreenshot(udid: string, format: "png" | "jpeg" = "png"): Promise<Buffer> {
  // simctl's stdout pipe mode (`-`) emits informational notes mixed with the
  // image bytes, which corrupts the output. Write to a tmp file instead.
  const { readFile, unlink } = await import("node:fs/promises");
  const { tmpdir } = await import("node:os");
  const { join } = await import("node:path");
  const tmpFile = join(tmpdir(), `claude-sim-${udid.slice(0, 8)}-${Date.now()}.${format}`);
  try {
    await execFileP("xcrun", ["simctl", "io", udid, "screenshot", `--type=${format}`, tmpFile], {
      maxBuffer: 16 * 1024 * 1024,
    });
    return await readFile(tmpFile);
  } finally {
    await unlink(tmpFile).catch(() => { /* ignore */ });
  }
}

export async function installApp(udid: string, appPath: string): Promise<void> {
  await execFileP("xcrun", ["simctl", "install", udid, appPath], { maxBuffer: 16 * 1024 * 1024 });
}

export async function launchApp(udid: string, bundleId: string): Promise<{ pid: number }> {
  const { stdout } = await execFileP("xcrun", ["simctl", "launch", udid, bundleId]);
  // Output format: "<bundleId>: <pid>"
  const m = stdout.match(/:\s*(\d+)/);
  return { pid: m ? Number(m[1]) : -1 };
}

File: apps/cockpit/src-tauri/src/simctl.rs
// Thin async wrappers around `xcrun simctl` for device lifecycle.

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use tokio::process::Command;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct SimDevice {
    pub udid: String,
    pub name: String,
    pub runtime: String,
    pub state: String,
}

pub async fn list_devices() -> Result<Vec<SimDevice>> {
    let output = Command::new("xcrun")
        .args(["simctl", "list", "devices", "--json", "available"])
        .output()
        .await
        .context("failed to run xcrun simctl list")?;

    if !output.status.success() {
        anyhow::bail!(
            "simctl list exit {}: {}",
            output.status,
            String::from_utf8_lossy(&output.stderr)
        );
    }

    let parsed: serde_json::Value = serde_json::from_slice(&output.stdout)
        .context("simctl list returned invalid JSON")?;

    let devices_map = parsed["devices"]
        .as_object()
        .context("simctl list: no 'devices' map")?;

    let mut result = Vec::new();
    for (runtime, devices) in devices_map {
        if let Some(arr) = devices.as_array() {
            for d in arr {
                if let (Some(udid), Some(name), Some(state)) = (
                    d["udid"].as_str(),
                    d["name"].as_str(),
                    d["state"].as_str(),
                ) {
                    result.push(SimDevice {
                        udid: udid.to_string(),
                        name: name.to_string(),
                        runtime: runtime.clone(),
                        state: state.to_string(),
                    });
                }
            }
        }
    }
    Ok(result)
}

/// Returns true if the device was already booted, false if we booted it.
pub async fn boot(udid: &str) -> Result<bool> {
    let devices = list_devices().await?;
    let target = devices
        .iter()
        .find(|d| d.udid == udid)
        .with_context(|| format!("device {} not found", udid))?;
    if target.state == "Booted" {
        return Ok(true);
    }

    let status = Command::new("xcrun")
        .args(["simctl", "boot", udid])
        .status()
        .await?;
    if !status.success() {
        anyhow::bail!("simctl boot {} failed: exit {}", udid, status);
    }

    // Wait for full boot
    let _ = Command::new("xcrun")
        .args(["simctl", "bootstatus", udid, "-b"])
        .status()
        .await?;
    Ok(false)
}

pub async fn shutdown(udid: &str) -> Result<()> {
    Command::new("xcrun")
        .args(["simctl", "shutdown", udid])
        .status()
        .await?;
    Ok(())
}
